question-service/controllers/questionController.js

```diff
@@ -0,0 +1,129 @@
+// controllers/controller.js
+const Question = require("../models/model");
+
+// Public routes
+exports.getAllQuestions = async (req, res) => {
+  try {
+    const data = await Question.find().sort({ title: 1 });
+    res.json(data);
+  } catch (error) {
+    res.status(500).json({ message: error.message });
+  }
+};
+
+exports.getQuestionById = async (req, res) => {
+  try {
+    const data = await Question.findById(req.params.id);
+    res.json(data);
+  } catch (error) {
+    res.status(500).json({ message: error.message });
+  }
+};
+
+exports.getRandomQuestion = async (req, res) => {
```

Comment:
I think it might be worth considering the MVCS pattern aka add an additional service layer to abstract away the logic here (and also in the other methods).

e.g. `questionController.js`:
```javascript
exports.getRandomQuestion = async (req, res) => {
  const { difficulty, categories } = req.query;

  try {
    const randomQuestion = await getRandomQuestionService(difficulty, categories);

    if (!randomQuestion) {
      return res.status(404).json({ message: "No questions found with the specified criteria" });
    }

    res.json(randomQuestion);
  } catch (error) {
    res.status(500).json({ error: "Error fetching random question" });
  }
};
```

e.g. `questionService.js`:
```javascript
exports.getRandomQuestionService = async (difficulty, categories) {
  // Filter questions based on difficulty and category
  const filter = {};
  if (difficulty) {
    filter.difficulty = difficulty;
  }
  if (categories) {
    filter.categories = categories;

  // Use your Mongoose model to query the database
  const filteredQuestions = await Question.find(filter);

  if (filteredQuestions.length === 0) {
    return null; // No questions found
  }

  // Select a random question from the filtered list
   const randomIndex = Math.floor(Math.random() * filteredQuestions.length);
  return filteredQuestions[randomIndex];
}
```

Reply:
Added a service layer